### PR TITLE
fix: preserve attachment sizes through upload and markdown restore

### PR DESCRIPTION
**Pull request by kristofferremback**
## Problem

- Upload pills showed the correct file size while the browser still owned the file, then flipped to `0 B` after upload because the backend trusted `multer-s3` size metadata that can stay zero.
- Inline attachment references also lost file metadata when content was serialized to markdown and parsed back, because the parser recreated attachment nodes with a hardcoded zero-byte size.

## Solution

- Recover attachment size from S3 in the backend whenever the upload middleware reports a non-positive size, and persist that corrected value.
- Encode attachment metadata into serialized attachment markdown and restore it on parse in the shared ProseMirror markdown layer.
- Treat legacy attachment markdown without metadata as unknown size (`null`) and render `Size unavailable` instead of a false `0 B`.

### How it works

1. The upload handler still receives file metadata from middleware, but `AttachmentService` now asks S3 for `ContentLength` when the reported size is invalid.
2. The shared ProseMirror serializer emits attachment links with a metadata title, for example `[report.pdf](attachment:attach_123 "threa-attachment:...")`.
3. The shared parser restores `filename`, `mimeType`, and `sizeBytes` from that metadata in both frontend and backend code paths.
4. Older markdown without metadata still parses, but attachment references keep `sizeBytes: null` rather than a fake zero.

### Key design decisions

**1. Fix the API contract at the backend boundary**

Use S3 object metadata as the source of truth so every client receives the correct attachment size. The alternative was to keep a frontend fallback, but that would leave the API wrong and repeat the bug in other clients.

**2. Keep attachment metadata inside the markdown representation**

Serialize attachment metadata into the attachment link title so draft and editor round-trips preserve fidelity without a separate persistence channel. The alternative was to rehydrate metadata from attachment lookups during parse, but parse runs in contexts that do not have attachment fetch access and must also work offline.

**3. Represent missing legacy metadata as `null`**

Older attachment markdown that never stored size information now yields `null` and renders `Size unavailable`. The alternative was to keep emitting `0`, but that encodes a lie and hides data loss.

## New files

| File | Purpose |
| --- | --- |
| `packages/prosemirror/src/markdown.test.ts` | Covers shared attachment markdown metadata serialization and legacy parsing behavior. |

## Modified files

| File | Change |
| --- | --- |
| `apps/backend/src/lib/storage/s3-client.ts` | Add S3 object size lookup via `HeadObject`. |
| `apps/backend/src/features/attachments/service.ts` | Recover and persist real attachment size when upload middleware reports zero. |
| `apps/backend/src/features/attachments/service.test.ts` | Add a regression test for zero-byte upload metadata recovery. |
| `apps/backend/evals/suites/companion/suite.ts` | Extend stub storage with `getObjectSize`. |
| `apps/backend/evals/suites/multimodal-vision/suite.ts` | Extend mock storage with `getObjectSize`. |
| `packages/types/src/prosemirror.ts` | Allow attachment reference `sizeBytes` to be `null`. |
| `packages/prosemirror/src/markdown.ts` | Preserve attachment metadata in markdown and restore it on parse. |
| `apps/frontend/src/components/editor/editor-markdown.ts` | Keep the local editor markdown implementation aligned with the shared parser. |
| `apps/frontend/src/components/editor/editor-markdown.test.ts` | Cover metadata-preserving round-trips and legacy null-size parsing. |
| `apps/frontend/src/components/editor/attachment-reference-extension.ts` | Parse and render null attachment sizes honestly in editor nodes. |
| `apps/frontend/src/components/editor/attachment-reference-view.tsx` | Render unknown attachment size as `Size unavailable`. |

## Deleted files

| File | Reason |
| --- | --- |
| None | No files deleted. |

## Test plan

- [x] `bun test src/features/attachments/service.test.ts`
- [x] `bun run typecheck` in `apps/backend`
- [x] `bunx vitest run src/hooks/use-attachments.test.ts`
- [x] `bunx vitest run src/components/editor/editor-markdown.test.ts`
- [x] `bun test ./packages/prosemirror/src/markdown.test.ts`
- [x] `bun run typecheck` in `packages/types`
- [x] `bun run typecheck` in `packages/prosemirror`
- [x] `bun run typecheck` in `apps/frontend`
- [ ] Run the browser upload flow end to end against the real UI

---

🤖 _PR by [Claude Code](https://claude.com/claude-code)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Attachment sizes are recovered from storage when upload reports zero; failures to determine size now surface as errors to avoid silent inconsistencies.

* **New Features**
  * Attachment metadata (filename, MIME type, size) is preserved in Markdown round-trips and editor views.
  * File size display now shows “Size unavailable” when size is missing.

### CODE REVIEW

**Comment by kristofferremback** [comment]
### Code review
**Confidence: 5/7** — Good
**Review models:** Orchestrator: claude-opus-4-6 | Reviewers: sonnet x3

Found 2 issues:

1. `s3-client.ts:49` — Silent `?? 0` fallback defeats the purpose of the fix (bug + CLAUDE.md [INV-11, §Architecture Boundaries]: "Fail loudly; no silent fallback defaults"). `getObjectSize` returns `response.ContentLength ?? 0`, silently falling back to `0` when `ContentLength` is absent from the HeadObject response. Since the service only calls `getObjectSize` when `params.sizeBytes` is already non-positive, a returned `0` silently persists the exact same wrong value this PR was introduced to fix. Should throw or return `null` so the caller can detect the failure.
   https://github.com/kristofferremback/threa/blob/f34a2af44b99da6c16801301417e90fd3f56b7e6/apps/backend/src/lib/storage/s3-client.ts#L48-L50

2. `editor-markdown.ts:11-64` / `markdown.ts:23-76` — Five attachment-metadata helpers duplicated verbatim across two files (CLAUDE.md [INV-35/INV-37, §Architecture Boundaries]: "Reuse existing helpers and abstractions instead of parallel implementations"). `ATTACHMENT_METADATA_PREFIX`, `escapeMarkdownLinkTitle`, `unescapeMarkdownLinkTitle`, `serializeAttachmentMetadata`, and `parseAttachmentMetadata` are copy-pasted identically into both `packages/prosemirror/src/markdown.ts` and `apps/frontend/src/components/editor/editor-markdown.ts`. The frontend already depends on `@threa/prosemirror` — these helpers should be exported from the shared package and imported in the frontend, not maintained as parallel implementations.
   https://github.com/kristofferremback/threa/blob/f34a2af44b99da6c16801301417e90fd3f56b7e6/apps/frontend/src/components/editor/editor-markdown.ts#L11-L64
   https://github.com/kristofferremback/threa/blob/f34a2af44b99da6c16801301417e90fd3f56b7e6/packages/prosemirror/src/markdown.ts#L23-L76

---
<details><summary>📐 Plan Adherence [CLEAN]</summary>All changes match the PR description. No missing corresponding changes detected.</details>
<details><summary>🔍 Bugs [1 issue]</summary>

- `s3-client.ts:49`: `getObjectSize` silently returns `0` via `?? 0` fallback when `ContentLength` is absent, re-introducing the same class of zero-byte size bug at a lower level. The service at `service.ts:43-46` only falls through to `getObjectSize` when `params.sizeBytes > 0` is false, so a returned `0` is persisted directly.

</details>
<details><summary>📋 CLAUDE.md Compliance [2 violations]</summary>

- **INV-11** (Fail loudly): `getObjectSize` uses `?? 0` as a silent fallback instead of throwing on missing `ContentLength`.
- **INV-35/INV-37** (Reuse helpers / No parallel implementations): Five new attachment metadata functions duplicated verbatim between the prosemirror package and the frontend editor-markdown module.

</details>
<details><summary>🏗️ Design [1 concern]</summary>

- The duplicated helpers mean any future change to the metadata serialization format (new fields, prefix change, escaping fix) must be applied in two places and kept in sync manually. The `@threa/prosemirror` package is the natural home for these, and the frontend already depends on it.

</details>
<details><summary>🔒 Security [CLEAN]</summary>No security issues found.</details>

🤖 Generated with unified-review automation
<sub>If this review was useful, react with 👍. Otherwise, react with 👎.</sub>

File: packages/prosemirror/src/markdown.ts [hunk 1]
```diff
@@ -116,7 +171,8 @@ function getNodeText(node: JSONContent): string {
     // Format: [Image #1](attachment:id) or [filename](attachment:id)
     const isImage = mimeType?.startsWith("image/")
     const displayText = isImage && imageIndex ? `Image #${imageIndex}` : filename
-    return `[${displayText}](attachment:${id})`
+    const metadata = serializeAttachmentMetadata(node.attrs)
+    return `[${displayText}](attachment:${id}${metadata})`
```
[the remark this answers is not in the document]
Comment (kristofferremback) [reply packages/prosemirror/src/markdown.ts]: Fixed. I moved the attachment markdown helpers into the shared ProseMirror package at `packages/prosemirror/src/attachment-markdown.ts` and imported them from the frontend editor markdown module instead of keeping parallel copies. Serialization now escapes `[` / `]` / `\\` in attachment labels, both parsers use the updated attachment regex and unescape on read, and I added regression coverage for escaped filenames in both the shared and frontend markdown tests.

🤖 _Response by [Claude Code](https://claude.com/claude-code)_
